text/0100-enhanced-pessimistic-lock-queueing.md

```diff
@@ -0,0 +1,223 @@
+
+- When the current command (maybe a commit or rollback) releases a lock and wakes up another acquire_pessimistic_lock command, we expect the latter command can success. To avoid another request arrives between the two command's execution and preempts the lock, we need to know on which key we are going to wake up a lock-waiting request, do not release its latch, and let the resumed command derive the latch instead of re-acquiring it.
+- We need to find blocked requests by key instead of by key hash, to avoid waking up wrong waiters and cause invalid execution of the command.
+- When waking up a lock-waiting request, we need the full parameters of that request and schedule a new command in scheduler to execute it.
+
+We are going to design a new lock waiting queue to meet our new requirement, as well as provide more extendability and possibility for future optimizations.
+
+The lock waiting queue can be accessed in the following places:
+
+- Pessimistic lock request encounters another lock: an entry should be enqueued to the waiting queue before the pessimistic lock command exits scheduler.
+- Releasing a lock (i.e. committing or rolling back a lock): it needs to check if there's any pessimistic lock request being blocked on this lock, wake one of them up if necessary. For requests that don't enable the new mechanism, the waking up procedure should be done before calling `engine.async_write` in `Scheduler::process_write` to return the response to client as early as possible. In this way, we can prevent the performance of the old locking mode from regressing.
+- A lock is successfully acquired by someone. It's possible that the state of a key transits from not-locked into locked while there are many pessimistic lock requests blocked on the key. For example, a transaction releases the lock and then one of the queueing transactions acquires the lock. In this case, the other queueing transactions is waiting for a different transaction from before, and the waiting information of the queueing transactions need to be updated in order to provide correct diagnostic information about lock waiting and do correct deadlock detection.
+- Delayed notify-all for requests that uses the old behavior. As we mentioned before, in the original version, after waking up one of the queueing requests, the other queueing requests on the same key will be woken up too after `wake-up-delay-duration`. This behavior should be kept in our new implementation for requests that choose not to use the new mode.
+
+Each queue will be a priority queue that pops the item with minimal `start_ts`, implemented by the std `BinaryHeap`. It's possible to extend the way to calculate the priority in the future, for example, considering its total wait time or the amount of transactions it's blocking.
+
+We put the queues in a concurrent hashmap indexed by key, and put the hashmap inner the `Scheduler`.
+
+#### Lazy-cleaning up
```

Comment:
Removal of an element is O(log n), whether lazily or not. Considering the problem mentioned, would it be better to just remove elements immediately when cancelling?

Reply:
When a request is canceled, it might not be at the head of the queue. It's possible to remove any element from a priority queue in O(log n) time, but it doesn't seem to be supported by the std `BinaryHeap`. I'm actually considering introducing a third-party implementation of priority queue or implement it by myself.